src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -95,6 +97,9 @@ bool ParentCacheObjectDispatch<I>::read(
         Context* on_finish = new LambdaContext([this](int ret) {
           m_connecting.store(false);
         });
+
+        locker.unlock();
+        std::unique_lock unique_locker{m_lock};
```

Comment:
@dillaman I am not sure I quite understand the purpose of re-acquiring the lock here.

Also the comment seems not entirely correct now. And I have a feeling with the lock it could be simplified. E.g. could we make m_connecting not atomic now? And could we avoid checking for `is_session_work` twice?

Reply:
It's because it's taking a shared lock at the top and a unique lock here. I could switch it to a regular mutex to avoid the atomic and avoid the race, but with the shared lock, it's possible that two threads pass the first `is_session_work` and the first one rebuilds the connection before the second one attempts to atomic exchange -- so you'd need to check again if the session is still dead or else you would restart a working session. Simplified it to just use a mutex for now since eventually it should also just use the existing asio io_context instead of starting its own.